[Blueprint] Enhance SQL execution safety

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOO6-10

The runSQL step is the most danger step in the blueprint import process, as it enables the execution of arbitrary SQL queries.

It's only accessible to users with the `manage_options` capability, but it's still a potential security risk if users import malicious blueprints without proper validation.

This PR enhances the security and reliability of SQL execution in the Blueprint's ImportRunSql class by implementing stricter validation and comprehensive test coverage.


### Key Changes

- Enhance SQL execution safety in the ImportRunSql class:
  - Add validation for SQL query types, only allow "INSERT", "UPDATE", "REPLACE INTO" queries
  - Implement protection against SQL injection patterns
  - Add safeguards for protected tables and user capabilities
  - Improve error handling and reporting
-  Refactor SQL query preparation in ExportWCTaxRates class for improved security
- Add comprehensive unit test coverage


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


### Test import

1. Enable blueprint feature `wp option set woocommerce_feature_blueprint_enabled yes`
2. Try importing an invalid SQL query; it should return an error  `wc blueprint import invalid-sql.json --show-messages=all`
3. Try importing a valid SQL query; it should return a success message  `wc blueprint import valid-sql.json --show-messages=all`
4. Act as a malicious user and try different malicious SQLs; it should return an error.

- [invalid-sql.json](https://github.com/user-attachments/files/19791393/invalid-sql.json)
- [valid-sql.json](https://github.com/user-attachments/files/19791394/valid-sql.json)


<img width="960" alt="Screenshot 2025-04-17 at 16 42 07" src="https://github.com/user-attachments/assets/baa79af4-7804-4f9d-9932-edb24b196900" />

### Test export tax

1. Create some tax rates
2. Run `wp wc blueprint export blueprint.json --steps=setWCTaxRates`
3. Confirm it generates the blueprint file with expected SQL statements


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Enhance SQL execution safety in ImportRunSql class

* Introduced validation checks for allowed SQL query types.
* Added mechanisms to detect SQL injection patterns and prevent unauthorized modifications to protected user tables and capabilities.
* Implemented SQL normalization to standardize query format.
* Enhanced error handling during SQL execution with transaction support.

chihsuan committed:
Add unit tests for ImportRunSql class

* Implemented comprehensive tests for SQL query processing, including valid INSERT, UPDATE, and REPLACE queries.
* Added tests to ensure invalid query types are correctly rejected and appropriate error messages are returned.
* Included tests for SQL normalization, handling of SQL injection patterns, and protection against modifications to user roles and capabilities.
* Ensured proper error handling for SQL execution failures and multiple statement queries.

chihsuan committed:
Add changelog

chihsuan committed:
Improve error handling in ImportRunSql class during SQL execution

* Updated error suppression to be enabled explicitly.
* Refactored error handling to store the last error in a variable for clarity.
* Enhanced error messages to provide more informative feedback on SQL execution failures.

chihsuan committed:
Refactor SQL handling in ImportRunSql class for improved security and clarity

* Removed the normalize_sql method and replaced it with direct SQL trimming.
* Added checks for suspicious SQL comments to enhance security against hidden malicious code.
* Updated related tests to validate detection of suspicious comments and ensure proper error handling for SQL queries.
* Simplified SQL validation logic by directly using trimmed SQL content in various checks.

chihsuan committed:
Fix tests

chihsuan committed:
Refactor SQL query preparation in ExportWCTaxRates class for improved security

* Updated the SQL query in generateSteps method to use prepared statements, enhancing security against SQL injection.
* Removed direct table name interpolation to ensure safer query execution.

chihsuan committed:
Update table name handling in ExportWCTaxRates class for improved SQL query preparation

* Modified the generateSteps method to prepend the table prefix directly to the table name variable, ensuring consistent and secure SQL query execution.
* This change enhances clarity and maintains the integrity of the SQL preparation process.

louwie17 commented on packages/php/blueprint/src/Importers/ImportRunSql.php:
```diff
@@ -79,4 +146,173 @@ public function check_step_capabilities( $schema ): bool {
+		// Quick check if there are any comments at all before running regex.
+		if ( strpos( $sql_content, '--' ) === false &&
+		strpos( $sql_content, '/*' ) === false &&
+		strpos( $sql_content, '#' ) === false ) {
```
Some formatting issues here

chihsuan replied:
Reformatted in d9ad44f2045eed0921f9be733d34bc8bae90d347

louwie17 commented on packages/php/blueprint/src/Importers/ImportRunSql.php:
```diff
@@ -79,4 +146,173 @@ public function check_step_capabilities( $schema ): bool {
+	 * @param string $sql_content The SQL query to check.
+	 * @return bool True if potential injection patterns found, false otherwise.
+	 */
+	private function contains_sql_injection_patterns( string $sql_content ): bool {
```
Curious where you got these patterns from and some of the other checks in this file?

chihsuan replied:
I used AI tools (ChatGPT, Claude, Cursor) to identify potential security issues and gather suggestions for enhancing the code's security measures.

louwie17 reviewed (approved):
I think the fact we are already limiting this by user type goes a long way, this also tested well and code looks good, although curious where you got these checks from?

There are a couple other things though that I am wondering about ( we could do this as follow up ):
- From looking at the code this would currently break for any WordPress instances that make use of table prefixes right? which brings me to the second suggestion.
- I know this mimics the Playground step, but I wonder if we should change the structure of the run sql step to limit it and mimic what insert, update functions that wpdb support https://developer.wordpress.org/reference/classes/wpdb/#update-rows ( it would also allow us to better support table prefixes )

chihsuan committed:
[Blueprint]  Add security restriction to only allow blueprint imports in Coming Soon mode (#57382)

* Implement import permission check for blueprint uploads

- Added a new API endpoint to check if blueprint imports are allowed based on site status.
- Integrated the import permission check into the file upload state machine.
- Updated the UI to display a notice when imports are disabled, guiding users to enable the "Coming Soon" mode or set a constant for live sites.
- Enhanced styling for error and notice components related to blueprint uploads.

This change improves user experience by providing clear feedback on import permissions.

* Add changelog

* Implement import permission check in the RestApi class

- Added a check in the import_step method to verify if blueprint imports are allowed.
- Returns an error message if imports are disabled, enhancing user feedback during the import process.

* Fix lints

* Add unit tests for blueprint import functionality

- Introduced tests to validate the import_step endpoint, including checks for file size limits and import permissions based on site status.
- Enhanced the setup and teardown methods for better resource management during tests.
- Ensured that appropriate error messages are returned when imports are disabled or when file size exceeds the limit, improving feedback for users.

* Fix lint

chihsuan committed:
Reformat

chihsuan commented:
> * From looking at the code this would currently break for any WordPress instances that make use of table prefixes right? which brings me to the second suggestion.

Yes, it would break for any WordPress instances that use table prefixes.

> * I know this mimics the Playground step, but I wonder if we should change the structure of the run sql step to limit it and mimic what insert, update functions that wpdb support https://developer.wordpress.org/reference/classes/wpdb/#update-rows ( it would also allow us to better support table prefixes )

This is a good idea! But I think using the same structure will help us use the same package across various projects. I'll check with the Playground team to see if they have any input. Thanks for the suggestion!

louwie17 commented:
> > * From looking at the code this would currently break for any WordPress instances that make use of table prefixes right? which brings me to the second suggestion.
> 
> Yes, it would break for any WordPress instances that use table prefixes.

Gotcha, that does make the `runSql` step not very useful in this context, as a lot of WP/WC sites use prefixes. I guess we will gather that from feedback the only way to really fix that is to change to format, as we shouldn't solve that with regex.

> This is a good idea! But I think using the same structure will help us use the same package across various projects. I'll check with the Playground team to see if they have any input. Thanks for the suggestion!

Thanks! I think the generic `runSql` makes a lot of sense in the Playground instance, but a bit less so for existing sites. It be nice if Playground could support both and we only support the more defined version.

louwie17 commented:
Btw feel free to merge this, I was just curious about the above questions, but they weren't blockers.

chihsuan committed:
Add changelog

chihsuan committed:
Merge branch 'trunk' into woo6-10-blueprints-security-audit-determine-the-scope-of-allowed-sql

chihsuan committed:
Merge branch 'trunk' into woo6-10-blueprints-security-audit-determine-the-scope-of-allowed-sql

chihsuan committed:
Remove unused imports

chihsuan committed:
Fix tests
